Update @wordpress packages

noisysocks opened the pull request:
Update packages to include these bug fixes from Gutenberg:

- https://github.com/WordPress/gutenberg/pull/36144
- https://github.com/WordPress/gutenberg/pull/36507
- https://github.com/WordPress/gutenberg/pull/36511
- https://github.com/WordPress/gutenberg/pull/36584
- https://github.com/WordPress/gutenberg/pull/36646
- https://github.com/WordPress/gutenberg/pull/36647
- https://github.com/WordPress/gutenberg/pull/36652
- https://github.com/WordPress/gutenberg/pull/36707
- https://github.com/WordPress/gutenberg/pull/36712
- https://github.com/WordPress/gutenberg/pull/36719
- https://github.com/WordPress/gutenberg/pull/36724
- https://github.com/WordPress/gutenberg/pull/36726
- https://github.com/WordPress/gutenberg/pull/36740
- https://github.com/WordPress/gutenberg/pull/36744
- https://github.com/WordPress/gutenberg/pull/36747
- https://github.com/WordPress/gutenberg/pull/36610
- https://github.com/WordPress/gutenberg/pull/36748
- https://github.com/WordPress/gutenberg/pull/36760
- https://github.com/WordPress/gutenberg/pull/36761
- https://github.com/WordPress/gutenberg/pull/36762
- https://github.com/WordPress/gutenberg/pull/36764
- https://github.com/WordPress/gutenberg/pull/36786
- https://github.com/WordPress/gutenberg/pull/36802
- https://github.com/WordPress/gutenberg/pull/36804
- https://github.com/WordPress/gutenberg/pull/36805
- https://github.com/WordPress/gutenberg/pull/36806
- https://github.com/WordPress/gutenberg/pull/36807
- https://github.com/WordPress/gutenberg/pull/36808
- https://github.com/WordPress/gutenberg/pull/36811
- https://github.com/WordPress/gutenberg/pull/36817
- https://github.com/WordPress/gutenberg/pull/36718
- https://github.com/WordPress/gutenberg/pull/36819
- https://github.com/WordPress/gutenberg/pull/36820
- https://github.com/WordPress/gutenberg/pull/36822
- https://github.com/WordPress/gutenberg/pull/36840
- https://github.com/WordPress/gutenberg/pull/36841
- https://github.com/WordPress/gutenberg/pull/36849
- https://github.com/WordPress/gutenberg/pull/36850
- https://github.com/WordPress/gutenberg/pull/36854
- https://github.com/WordPress/gutenberg/pull/36877
- https://github.com/WordPress/gutenberg/pull/36878
- https://github.com/WordPress/gutenberg/pull/36815
- https://github.com/WordPress/gutenberg/pull/36879
- https://github.com/WordPress/gutenberg/pull/36880
- https://github.com/WordPress/gutenberg/pull/36881
- https://github.com/WordPress/gutenberg/pull/36893
- https://github.com/WordPress/gutenberg/pull/36907
- https://github.com/WordPress/gutenberg/pull/36911
- https://github.com/WordPress/gutenberg/pull/36913
- https://github.com/WordPress/gutenberg/pull/36918
- https://github.com/WordPress/gutenberg/pull/36921
- https://github.com/WordPress/gutenberg/pull/36684
- https://github.com/WordPress/gutenberg/pull/36922
- https://github.com/WordPress/gutenberg/pull/36739
- https://github.com/WordPress/gutenberg/pull/36876
- https://github.com/WordPress/gutenberg/pull/36931
- https://github.com/WordPress/gutenberg/pull/36763
- https://github.com/WordPress/gutenberg/pull/36896

Trac ticket: https://core.trac.wordpress.org/ticket/54487

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

talldan committed:
Add support to post types

talldan committed:
Add author support to templates REST API controller

talldan committed:
Update unit tests

talldan committed:
Fix linting issue

talldan committed:
Add origin field to templates

talldan committed:
Handle errors from `prepare_item_for_database` in REST templates controller

talldan committed:
Update tests to support origin

talldan committed:
Address review feedback

talldan committed:
Linting fix

talldan committed:
Remove empty lines

noisysocks committed:
Changes from lib/class-wp-theme-json-gutenberg.php

noisysocks committed:
Changes from lib/class-wp-theme-json-resolver-gutenberg.php

noisysocks committed:
Changes from lib/class-gutenberg-rest-global-styles-controller.php

noisysocks committed:
Changes from lib/compat/wordpress-5.9/class-gutenberg-rest-templates-controller.php

noisysocks committed:
Changes from lib/compat/wordpress-5.9/get-global-styles-and-settings.php

noisysocks committed:
Changes from lib/compat/wordpress-5.9/rest-active-global-styles.php

noisysocks committed:
Changes from lib/full-site-editing/edit-site-page.php

noisysocks committed:
Changes from lib/global-styles.php

noisysocks committed:
Changes from phpunit/class-gutenberg-rest-template-controller-test.php

noisysocks committed:
Changes from phpunit/class-wp-theme-json-resolver-test.php

noisysocks committed:
Changes from phpunit/class-wp-theme-json-test.php

noisysocks committed:
Update @wordpress packages

noisysocks commented:
I've yet to test this. It's possible that some or all of these need to be committed first:

- https://core.trac.wordpress.org/ticket/54506
- https://core.trac.wordpress.org/ticket/54448
- https://github.com/WordPress/wordpress-develop/pull/1937

noisysocks committed:
Move call to maybe_opt_in_into_settings() to right place

noisysocks committed:
Use WP class names

noisysocks commented:
I've merged https://github.com/WordPress/wordpress-develop/pull/1937 into this branch.

noisysocks commented:
@Mamaduka: Could you please fix the `is_custom` PHP unit test failures here? I probably messed up porting the PHP changes in https://github.com/WordPress/gutenberg/pull/36911. It's confusing because `Tests_REST_WpRestTemplatesController` seems to have some pretty significant differences between Core and the plugin.

@adamziel: Do you know why `Tests_REST_WpRestTemplatesController::test_get_item_works_with_a_single_slash` from https://github.com/WordPress/gutenberg/pull/36881 fails here?

Mamaduka commented:
@noisysocks we should probably sync `Tests_REST_WpRestTemplatesController` tests back into plugin.

Mamaduka committed:
Fix 'is_custom' assertions in PHPUnit tests

adamziel commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
```diff
@@ -143,11 +146,76 @@ public function test_get_item() {
+	/**
+	 * @ticket 54507
+	 * @dataProvider get_template_endpoint_urls
+	 */
+	public function test_get_item_works_with_a_single_slash( $endpoint_url ) {
+		wp_set_current_user( self::$admin_id );
+		$request  = new WP_REST_Request( 'GET', $endpoint_url );
+		$response = rest_get_server()->dispatch( $request );
+
+		$data = $response->get_data();
+		unset( $data['content'] );
+		unset( $data['_links'] );
+
+		$this->assertEquals(
+			array(
+				'id'             => 'tt1-blocks//index',
+				'theme'          => 'tt1-blocks',
+				'slug'           => 'index',
+				'title'          => array(
+					'raw'      => 'Index',
+					'rendered' => 'Index',
+				),
+				'description'    => 'The default template used when no other template is available. This is a required template in WordPress.',
+				'status'         => 'publish',
+				'source'         => 'theme',
+				'type'           => 'wp_template',
+				'wp_id'          => null,
+				'has_theme_file' => true,
+				'is_custom'      => false,
+			),
+			$data
+		);
+	}
+
+	public function get_template_endpoint_urls() {
+		return array(
+			array( '/wp/v2/templates/tt1-blocks/index' ),
+			array( '/wp/v2/templates/tt1-blocks//index' ),
+		);
+	}
```
This test uses different template names from the Gutenberg version. Ideally we'd get both in sync with each other. The following change will get this PR across the line:

```suggestion	/**
	 * @ticket 54507
	 * @dataProvider get_template_endpoint_urls
	 */
	public function test_get_item_works_with_a_single_slash( $endpoint_url ) {
		wp_set_current_user( self::$admin_id );
		$request  = new WP_REST_Request( 'GET', $endpoint_url );
		$response = rest_get_server()->dispatch( $request );

		$data = $response->get_data();
		unset( $data['content'] );
		unset( $data['_links'] );

		$this->assertSame(
			array(
				'id'             => 'default//my_template',
				'theme'          => 'default',
				'slug'           => 'my_template',
				'source'         => 'custom',
				'origin'         => null,
				'type'           => 'wp_template',
				'description'    => 'Description of my template.',
				'title'          => array(
					'raw'      => 'My Template',
					'rendered' => 'My Template',
				),
				'status'         => 'publish',
				'wp_id'          => self::$post->ID,
				'has_theme_file' => false,
				'is_custom'      => true,
				'author'         => 0,
			),
			$data
		);
	}

	public function get_template_endpoint_urls() {
		return array(
			array( '/wp/v2/templates/default/my_template' ),
			array( '/wp/v2/templates/default//my_template' ),
		);
	} ```

noisysocks committed:
Update tests/phpunit/tests/rest-api/wpRestTemplatesController.php

Co-authored-by: Adam Zielinski <adam@adamziel.com>

adamziel commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
```diff
@@ -143,11 +146,77 @@ public function test_get_item() {
 		);
 	}
 
+	 * @ticket 54507
```
```suggestion
	/**
	 * @ticket 54507
```

adamziel commented on tests/phpunit/tests/rest-api/wpRestTemplatesController.php:
````diff
@@ -143,11 +146,77 @@ public function test_get_item() {
+			array( '/wp/v2/templates/default/my_template' ),
+			array( '/wp/v2/templates/default//my_template' ),
+		);
+	} ```
````
```suggestion
	}
```

noisysocks committed:
Update tests/phpunit/tests/rest-api/wpRestTemplatesController.php

Co-authored-by: Adam Zielinski <adam@adamziel.com>

noisysocks committed:
Update tests/phpunit/tests/rest-api/wpRestTemplatesController.php

Co-authored-by: Adam Zielinski <adam@adamziel.com>

hellofromtonya commented:
Noting this Trac ticket https://core.trac.wordpress.org/ticket/54532 and Gutenberg issue https://github.com/WordPress/gutenberg/issues/36962 for consideration _before_ Beta 1.

oandregal commented:
~Robert, JFYI, in case is not merged by the time you read this. https://github.com/WordPress/wordpress-develop/pull/1971 would be nice to have for Beta 1 (Nov 30) as it adds a couple of missing `theme.json` flags missing from the the last backport.~

Update: I've noticed we were missing another flag (`appearanceTools`) in this backport, so I've prepared a single PR to add them all at https://github.com/WordPress/wordpress-develop/pull/1973

oandregal committed:
Sort keys alphabetically so it is easier to compare

oandregal committed:
Add missing appearanceTools flag

oandregal commented:
Hi, looking at the `class-theme-json.php` file I've found a thing to fix and a couple to improve. Apparently, I can't prepare a PR on top of this one, so I thought it'd be best in this case to share a patch to be applied by `git apply <file.patch>`:

The contents of `file.patch` are:

```patch
diff --git a/src/wp-includes/class-wp-theme-json.php b/src/wp-includes/class-wp-theme-json.php
index 4cdb44e457..635f670cff 100644
--- a/src/wp-includes/class-wp-theme-json.php
+++ b/src/wp-includes/class-wp-theme-json.php
@@ -75,17 +75,28 @@ class WP_Theme_JSON {
 	 * This contains the necessary metadata to process them:
 	 *
 	 * - path       => where to find the preset within the settings section
-+	 * - override   => whether a theme preset with the same slug as a default preset
-+	 *                 can override it
+	 * - override   => whether a theme preset with the same slug as a default preset
+	 *                 can override it
 	 * - value_key  => the key that represents the value
-	 * - value_func => the callback to render the value (either value_key or value_func should be present)
+	 * - value_func => optionally, instead of value_key, a function to generate
+	 *                 the value that takes a preset as an argument
+	 *                 (either value_key or value_func should be present)
 	 * - css_vars   => template string to use in generating the CSS Custom Property.
 	 *                 Example output: "--wp--preset--duotone--blue: <value>" will generate as many CSS Custom Properties as presets defined
 	 *                 substituting the $slug for the slug's value for each preset value.
-	 * - classes    => array containing a structure with the classes to generate for the presets.
-	 *                 Each key is a template string to resolve similarly to the css_vars and each value is the CSS property to use for that class.
-	 *                 Example output: ".has-blue-color { color: <value> }"
-	 * - properties  => a list of CSS properties to be used by kses to check the preset value is safe.
+	 * - classes    => array containing a structure with the classes to
+	 *                 generate for the presets, where for each array item
+	 *                 the key is the class name and the value the property name.
+	 *                 The "$slug" substring will be replaced by the slug of each preset.
+	 *                 For example:
+	 *                 'classes' => array(
+	 *                   '.has-$slug-color'            => 'color',
+	 *                   '.has-$slug-background-color' => 'background-color',
+	 *                   '.has-$slug-border-color'     => 'border-color',
+	 *                 )
+	 * - properties => array of CSS properties to be used by kses to
+	 *                 validate the content of each preset
+	 *                 by means of the remove_insecure_properties method.
 	 *
 	 * @since 5.8.0
 	 * @since 5.9.0 Added new presets and simplified the metadata structure.
@@ -305,6 +316,8 @@ class WP_Theme_JSON {
 	);
 
 	/**
+	 * The valid elements that can be found under styles.
+	 *
 	 * @since 5.8.0
 	 * @var string[]
 	 */
@@ -479,6 +492,7 @@ class WP_Theme_JSON {
 
 		return $output;
 	}
+
 	/**
 	 * Returns the metadata for each block.
 	 *
```

oandregal commented on src/wp-includes/class-wp-theme-json.php:
```diff
@@ -75,6 +75,8 @@ class WP_Theme_JSON {
 	 * This contains the necessary metadata to process them:
 	 *
 	 * - path       => where to find the preset within the settings section
++	 * - override   => whether a theme preset with the same slug as a default preset
```
There're a couple of extra `+` here. I've sent a patch to fix that and other minor changes at https://github.com/WordPress/wordpress-develop/pull/1967#issuecomment-981965618

oandregal replied:
I've also prepared this Gutenberg PR https://github.com/WordPress/gutenberg/pull/36973 to make sure that the order of constants, methods, etc. as well as the PHPDoc comments are the same between the Gutenberg's `lib/class-wp-theme-json-gutenberg.php` and WordPress core `wp-includes/class-wp-theme-json.php`.

This way the backports are way easier and require minimal manual intervention.

oandregal commented:
Got to review the `class-wp-theme-json-resolver.php` class as well. I've only noticed minor changes to the comments:

```patch
diff --git a/src/wp-includes/class-wp-theme-json-resolver.php b/src/wp-includes/class-wp-theme-json-resolver.php
index 6a7e6a5bf9..659365b6ef 100644
--- a/src/wp-includes/class-wp-theme-json-resolver.php
+++ b/src/wp-includes/class-wp-theme-json-resolver.php
@@ -230,7 +230,6 @@ class WP_Theme_JSON_Resolver {
         * @param array    $post_status_filter Filter Optional. custom post type by
         *                                     post status.  ['publish'] by default,
         *                                     so it only fetches published posts.
-        *
         * @return array Custom Post Type for the user's origin config.
         */
        public static function get_user_data_from_custom_post_type( $theme, $should_create_cpt = false, $post_status_filter = array( 'publish' ) ) {
@@ -333,8 +332,8 @@ class WP_Theme_JSON_Resolver {
 
        /**
         * There are three sources of data (origins) for a site:
-        * default, theme, and user. The user's has higher priority
-        * than the theme's, and the theme's higher than core's.
+        * default, theme, and custom. The custom's has higher priority
+        * than the theme's, and the theme's higher than defaults's.
         *
         * Unlike the getters {@link get_core_data},
         * {@link get_theme_data}, and {@link get_user_data},
```

I'm making some changes to the equivalent Gutenberg class to make the backports easier at https://github.com/WordPress/gutenberg/pull/36974

spacedmonkey commented:
We need to ensure that https://github.com/WordPress/gutenberg/pull/36981 is merged and backported as well.

noisysocks commented:
Thanks @oandregal for the review and fixes. I'm going to apply those patches and commit this patch, and _then_ follow up with a second commit to address https://github.com/WordPress/gutenberg/issues/36962 and https://github.com/WordPress/gutenberg/pull/36981. That way if I run into roadblocks we'll still have something in for beta 1 😀

hellofromtonya commented:
@noisysocks need to make sure that the `blocks/navigation-area` folder and file are also removed from Core on package update.

noisysocks committed:
Apply class-theme-json.php patch

noisysocks committed:
Apply class-wp-theme-json-resolver.php patch

noisysocks commented:
> @noisysocks need to make sure that the `blocks/navigation-area` folder and file are also removed from Core on package update.

This patch doesn't modify `navigation-area`. I've rebased this branch though so that that folder doesn't exist in this branch.

noisysocks commented:
Gave this a smoke test locally and nothing seems amiss. Committing!
